sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -1882,3 +1882,141 @@ case class ArrayRepeat(left: Expression, right: Expression)
   }
 
 }
+
+/**
+ * Removes duplicate values from the array.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(array) - Removes duplicate values from the array.",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(1, 2, 3, null, 3));
+       [1,2,3,null]
+  """, since = "2.4.0")
+case class ArrayDistinct(child: Expression)
+  extends UnaryExpression with ExpectsInputTypes {
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(ArrayType)
+
+  override def dataType: DataType = child.dataType
+
+  lazy val elementType: DataType = dataType.asInstanceOf[ArrayType].elementType
+
+  override def nullSafeEval(array: Any): Any = {
+    val elementType = child.dataType.asInstanceOf[ArrayType].elementType
+    val data = array.asInstanceOf[ArrayData].toArray[AnyRef](elementType).distinct
+    new GenericArrayData(data.asInstanceOf[Array[Any]])
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    nullSafeCodeGen(ctx, ev, (array) => {
+      val i = ctx.freshName("i")
+      val j = ctx.freshName("j")
+      val hs = ctx.freshName("hs")
+      val distinctArrayLen = ctx.freshName("distinctArrayLen")
+      val getValue = CodeGenerator.getValue(array, elementType, i)
+      val openHashSet = classOf[OpenHashSet[_]].getName
+      val classTag = s"scala.reflect.ClassTag$$.MODULE$$.Object()"
+      s"""
+         |int $distinctArrayLen = 0;
+         |$openHashSet $hs = new $openHashSet($classTag);
+         |for (int $i = 0; $i < $array.numElements(); $i++) {
+         |  if ($array.isNullAt($i)) {
+         |    int $j;
```

Comment:
IIUC, do we need to execute this then-block only for the first null element in `$array`?

Reply:
@kiszk Thanks for your comment. I have changed the code.